CRAMcodecs.tex

```diff
@@ -0,0 +1,3183 @@
+% \hline
+% ploc   & 9  & With $ptab$ performing pos/32 function.\\
+% \hline
+% dloc   & 12 & With $dtab$ performing delta/4 function.\\
+% \hline
+% rloc   & 15 & 1 bit, iff $do\_read2$ flag is set\\
+% \hline
+% \end{tabular}
+%
+% FIXME: our worked example should include actual bytes for qmap, ptab
+% and dtab too.
+
+
+\textsc{FQZDecodeParams} below describes the pseudocode for reading
+the parameter block.
+
+\begin{algorithmic}[1]
+\Procedure{FQZDecodeParams}{}
+  \State $vers \gets $\Call{ReadUint8}{}
+  \If{$vers \ne 5$}
+    \State ERROR
+  \EndIf
+  \State $gflags \gets$ \Call{ReadUint8}{}
+  \If{$gflags \bitand 1$} \Comment{$multi\_param$}
+    \State $nparam \gets$ \Call{ReadUint8}{}
+    \State $max\_sel \gets nparam$
+  \Else
+    \State $nparam \gets 1$
+    \State $max\_sel \gets 0$
+  \EndIf
+  \If{$gflags \bitand 2$} \Comment{$have\_stab$}
+    \State $max\_sel \gets$ \Call{ReadUint8}{}
+    \State $stab \gets$ \Call{ReadArray}{$256$}
+  \EndIf
+  \State $max\_sym \gets 0$
+  \For{$p \gets 0$ to $nparam-1$}
+    \State $param_p \gets$ \Call{FQZDecodeSingleParam}{}
+    \If {$max\_sym < param_p.max\_sym$}
+      \State $max\_sym \gets param_p.max\_sym$ \Comment{Maximum across all param sets}
+    \EndIf
+  \EndFor
+\EndProcedure
+\end{algorithmic}
+
+\begin{algorithmic}[1]
+\Function{FQZDecodeSingleParam}{}
+  \settowidth{\maxwidth}{p.have\_qtab\ }
+  \State \algalign{p.context}{\gets} \Call{ReadUint16}{}
+  \State \algalign{p.flags}{\gets} \Call{ReadUint8}{}
+%  \State \algalign{have\_qtab}{\gets}    $p.flags\bitand 128$
+%  \State \algalign{have\_dtab}{\gets}    $p.flags\bitand 64$
+%  \State \algalign{have\_ptab}{\gets}    $p.flags\bitand 16$
+%  \State \algalign{do\_rev}{\gets}    $flags\bitand 16$
+%  \State \algalign{do\_sel}{\gets} $flags\bitand 8$
+%  \State \algalign{do\_len}{\gets}    $flags\bitand 4$
+%  \State \algalign{do\_dedup}{\gets}  $flags\bitand 2$
+%  \State \algalign{have\_qmap}{\gets}   $flags\bitand 1$
+  \State \algalign{p.max\_sym}{\gets} \Call{ReadUint8}{}
+  \State \algalign{p.first\_len}{\gets} $1$
+
+  \settowidth{\maxwidth}{p.qshift\ }
+  \State \algalign{x}{\gets} \Call{ReadUint8}{}
+  \State \algalign{p.qbits}{\gets} $x \bdiv 16$
+  \State \algalign{p.qshift}{\gets} $x \bmod 16$
+  \State \algalign{x}{\gets} \Call{ReadUint8}{}
+  \State \algalign{p.qloc}{\gets} $x \bdiv 16$
+  \State \algalign{p.sloc}{\gets} $x \bmod 16$
+  \State \algalign{x}{\gets} \Call{ReadUint8}{}
+  \State \algalign{p.ploc}{\gets} $x \bdiv 16$
+  \State \algalign{p.dloc}{\gets} $x \bmod 16$
+
+  \If{$p.flags\bitand 1$} \Comment{Have qmap}
+    \For{$i \gets 0$ to $p.max\_sym-1$}
+      \State $p.qmap_i \gets$ \Call{ReadUint8}{}
+    \EndFor
+  \EndIf
+
+  \If{$p.flags\bitand 128$} \Comment{Have qtab}
+    \State $p.qtab \gets$ \Call{ReadArray}{$256$}
+  \Else
+    \For{$i \gets 0$ to $256$}
+      \State $p.qtab_i \gets i$
+    \EndFor
+  \EndIf
+
+  \If{$p.flags\bitand 16$} \Comment{Have ptab}
```

Comment:
```suggestion
  \If{$p.flags\bitand 32$} \Comment{Have ptab}
```

Reply:
I noticed the dtab field was misspelt following that too.  Both fixed.